engine/hatchery/vsphere/ip.go

```diff
@@ -2,71 +2,59 @@ package vsphere
 
 import (
 	"context"
-	"fmt"
-	"math/rand"
+	"errors"
 	"time"
+
+	"github.com/ovh/cds/sdk"
 )
 
 // for each ip in the range, look for the first free ones
 func (h *HatcheryVSphere) findAvailableIP(ctx context.Context, workerName string) (string, error) {
-	srvs := h.getServers(ctx)
+	h.IpAddressesMutex.Lock()
+	defer h.IpAddressesMutex.Unlock()
+
+	srvs := h.getVirtualMachines(ctx)
 
-	ipsInfos.mu.Lock()
-	defer ipsInfos.mu.Unlock()
-	for ip, infos := range ipsInfos.ips {
-		// ip used less than 10s
-		// 15s max to display worker name on nova list after call create
-		if time.Since(infos.dateLastBooked) < 15*time.Second {
+	var usedIPAddresses = make(map[string]struct{}, len(srvs))
+	for _, s := range srvs {
+		if s.Guest == nil {
 			continue
 		}
-		found := false
-		for _, srv := range srvs {
-			if infos.workerName == srv.Name {
-				found = true
+		for _, n := range s.Guest.Net {
+			for _, ip := range n.IpAddress {
+				usedIPAddresses[ip] = struct{}{}
+				// If the IP Address is know as a reservedIPAddress, remove it
+				if sdk.IsInArray(ip, h.reservedIPAddresses) {
+					h.reservedIPAddresses = sdk.DeleteFromArray(h.reservedIPAddresses, ip)
+				}
 			}
 		}
-		if !found {
-			infos.workerName = ""
-			ipsInfos.ips[ip] = infos
-		}
 	}
 
-	freeIP := []string{}
-	for ip := range ipsInfos.ips {
-		free := true
-		if ipsInfos.ips[ip].workerName != "" {
-			continue // ip already used by a worker
-		}
-	serverLoop:
-		for _, s := range srvs {
-			if s.Guest == nil {
-				continue
-			}
-			for _, n := range s.Guest.Net {
-				for _, vmIP := range n.IpAddress {
-					if vmIP == ip {
-						free = false
-						break serverLoop
-					}
-				}
-
-			}
-		}
-		if free {
-			freeIP = append(freeIP, ip)
+	for _, ip := range h.availableIPAddresses {
+		_, isUsed := usedIPAddresses[ip]
+		isReserved := sdk.IsInArray(ip, h.reservedIPAddresses)
+		if !isUsed && !isReserved {
+			return ip, nil
 		}
 	}
+	return "", sdk.WithStack(errors.New("no IP address available"))
+}
 
-	if len(freeIP) == 0 {
-		return "", fmt.Errorf("no IP left")
+func (h *HatcheryVSphere) reserveIPAddress(ctx context.Context, ip string) error {
+	h.IpAddressesMutex.Lock()
+	if sdk.IsInArray(ip, h.reservedIPAddresses) {
+		return sdk.WithStack(errors.New("address already reserved"))
 	}
+	h.reservedIPAddresses = append(h.reservedIPAddresses, ip)
+	h.IpAddressesMutex.Unlock()
 
-	ipToBook := freeIP[rand.Intn(len(freeIP))]
-	infos := ipInfos{
-		workerName:     workerName,
-		dateLastBooked: time.Now(),
-	}
-	ipsInfos.ips[ipToBook] = infos
+	go func() {
```

Comment:
Can we wait syncro, its strange to have this kind of code that can change the hatchery at any moment, no ?

Reply:
same weird cache implementation in all the vsphere hatchery